libs/hdf-converters/src/sbom-mapper.ts

```diff
@@ -0,0 +1,482 @@
+]);
+
+function formatCWETags(input: number[], addPrefix = true): string[] {
+  const stringifiedCWE: string[] = [];
+  for (const cwe of input) {
+    const cweTag = addPrefix ? `CWE-${cwe}` : `${cwe}`;
+    stringifiedCWE.push(cweTag);
+  }
+  return stringifiedCWE;
+}
+
+function getNISTTags(input: number[]): string[] {
+  return CWE_NIST_MAPPING.nistFilter(
+    formatCWETags(input, false),
+    DEFAULT_NIST_TAG
+  );
+}
+
+// A single SBOM vulnerability can contain multiple security ratings
+// Average any existing ratings and then pass to `impact`
+function aggregateImpact(ratings: Record<string, unknown>[]): number {
+  let impact = 0;
+  for (const rating of ratings) {
+    // Prefer to use CVSS-based `score` field when possible
+    if (_.has(rating, 'score') && _.get(rating, 'method') == 'CVSSv31') {
+      impact += (rating as {score: number}).score;
+    } else {
+      // Else interpret it from `severity` field
+      const severity = IMPACT_MAPPING.get(
+        (rating as {severity: string}).severity.toLowerCase()
+      );
+      if (severity) {
+        impact += severity * 10;
+      }
+    }
+  }
+  // Round up aggregate impact to the 2nd decimal place
+  return Math.ceil((impact / ratings.length) * 10) / 100;
+}
+
+export class SBOMResults {
+  data: Record<string, unknown>;
+  withRaw: boolean;
+  constructor(SBOMJson: string, withRaw = false) {
+    this.data = JSON.parse(SBOMJson);
+    this.withRaw = withRaw;
+
+    if (_.has(this.data, 'components')) {
+      // In-place manipulations on ingested SBOM data
+      this.flattenComponents(this.data);
+      if (_.has(this.data, 'vulnerabilities')) {
+        this.generateIntermediary(this.data);
+      }
+    } else if (_.has(this.data, 'vulnerabilities')) {
+      // Back up operations in case we ingest VEX data instead
+      this.formatVEX(this.data);
+    }
+  }
+
+  // Flatten any arbitrarily nested components list
+  flattenComponents(data: Record<string, unknown>): void {
+    // Look through every component at the top level of the list
+    for (const component of data.components as Record<string, unknown>[]) {
+      // Identify if subcomponents exist
+      if (_.has(component, 'components')) {
+        // If so, pull out the subcomponents and push them to end of top level component list for further flattening
+        for (const subcomponent of component.components as Record<
+          string,
+          unknown
+        >[]) {
+          (data.components as Record<string, unknown>[]).push(subcomponent);
+        }
+        delete component.components;
+      }
+    }
+  }
+
+  /*
+  Copy all components that are affected by a vulnerability and place them under that corresponding vulnerability
+  In-place operation on `vulnerabilities` structure but will not affect `components` structure
+
+  Should result in the following general structure:
+  {
+    components: [...],
+    vulnerabilities: [
+      vulnerability: {
+        affectedComponents: [
```

Comment:
could you clarify that the affected components section here is what is added by this function

Reply:
Addressed.